### ISSUE TITLE
EfficientDet Training loss=nan 

### ISSUE THREAD

**Issue by etatbak** [tensorflow/models#9151]
I'm training my custom model with EfficientDet D0 but after 700 steps I am getting loss as nan value. Is there someone who has the same problem?
TensorFlow 2.3.0 with GTX 1060 10.1 CUDA

Here my training overview:
![image](https://user-images.githubusercontent.com/46399056/91206063-d1914500-e706-11ea-926f-39c5c9fb1bb4.png)

I am using default config file parameters. Image size 512x512 and batch_size 1

**Comment by TolgaBkm** [comment]
I have the same problem with EfficientDet D4.
![D4](https://user-images.githubusercontent.com/57003251/91416865-a61e6f80-e858-11ea-8aa9-4a3076a19a1e.PNG)

**Comment by TannerGilbert** [comment]
One of my viewers had the same problem. Changing the batch size from 1 to 4 solved the problem

**Comment by koolvn** [comment]
I also have the same problem. I've tried d1 and d2. 
![image](https://user-images.githubusercontent.com/48523533/91523732-ed6b3580-e916-11ea-8cd8-cd1a94ceacac.png)
Also, it seems that network is not learning at all.. I have a custom .tfrecord dataset. The config file is attached.

[pipeline_d2.config.txt](https://github.com/tensorflow/models/files/5140141/pipeline_d2.config.txt)

**Comment by TolgaBkm** [comment]
I have also tried to increase the batch size to 4. However, I still got loss=NaN around 50k steps of training.

**Comment by geeheim** [comment]
Try to reduce the learning rate, this should solve the problem. The models are usually trained with a large batchsize, eg.: 128 for efficient_det d3, when adjusting this parameter (eg: 128 -> 4) makes sure to also change the learning rate, since your gradients are a lot noisier. I would divide the learning rate by the same factor.

**Comment by koolvn** [comment]
> Try to reduce the learning rate, this should solve the problem. The models are usually trained with a large batchsize, eg.: 128 for efficient_det d3, when adjusting this parameter (eg: 128 -> 4) makes sure to also change the learning rate, since your gradients are a lot noisier. I would divide the learning rate by the same factor.

Thank you for an advice! I've never thought that learning rate could cause a problem like that. 
Now I'm training EfficientDet D1 on GTX 1080 Ti with batch size = 6 and LR = 0.003 with warmup. It looks promising.
![image](https://user-images.githubusercontent.com/48523533/91583351-88412f80-e96a-11ea-9a63-876b19424fe3.png)

Here's [my pipeline config](https://github.com/tensorflow/models/files/5142814/pipeline_d1.config.txt) in case anyone ever needs it.

**Comment by aafaqin** [comment]
> > Try to reduce the learning rate, this should solve the problem. The models are usually trained with a large batchsize, eg.: 128 for efficient_det d3, when adjusting this parameter (eg: 128 -> 4) makes sure to also change the learning rate, since your gradients are a lot noisier. I would divide the learning rate by the same factor.
> 
> @greeheim Thank you for an advice! I've never thought that learning rate could cause a problem like that.
> Now I'm training EfficientDet D1 on GTX 1080 Ti with batch size = 6 and LR = 0.003 with warmup. It looks promising.
> ![image](https://user-images.githubusercontent.com/48523533/91583351-88412f80-e96a-11ea-9a63-876b19424fe3.png)
> 
> Here's [my pipeline config](https://github.com/tensorflow/models/files/5142814/pipeline_d1.config.txt) in case anyone ever needs it.
I am getting a similar issue right now tried to reduce the learning rate of the model and training seems working  well.
Those blue lines are on tensorboard logs record, is it testing metrics? how do i get it while training?
I am only able to get training metrics am using COCO metrics.
also on evaluation am able to get only single point after training. how to get this graph

**Comment by koolvn** [comment]
@aafaqin If you're using tf2 then you should run a separate evaluation process to get those blue lines in tensorboard.
Use smth like this
```
# eval
python models/research/object_detection/model_main_tf2.py --pipeline_config_path=$PIPELINE_CONFIG --model_dir=$MODEL_DIR --checkpoint_dir=$MODEL_DIR --alsologtostderr
```

**Comment by adityap27** [comment]
> @aafaqin If you're using tf2 then you should run a separate evaluation process to get those blue lines in tensorboard.
> Use smth like this
> 
> ```
> # eval
> python models/research/object_detection/model_main_tf2.py --pipeline_config_path=$PIPELINE_CONFIG --model_dir=$MODEL_DIR --checkpoint_dir=$MODEL_DIR --alsologtostderr
> ```

@koolvn I am still not able to see those blue lines....is it after some steps or epochs or so ? also for tensorboard which logdir to use ? /train or /eval ....I can see 2 logs dirs.

**Comment by koolvn** [comment]
@aafaqin are you running evaluation task in parallel with training?
Yes, it'll make evaluate every 1000 steps as far as I can remember
Also you have to specify evaluation dataset in your config file
![image](https://user-images.githubusercontent.com/48523533/99932835-13b83a00-2d7b-11eb-81b7-f26a73dc97de.png)

**Comment by adityap27** [comment]
> @aafaqin are you running evaluation task in parallel with training?
> Yes, it'll make evaluate every 1000 steps as far as I can remember
> Also you have to specify evaluation dataset in your config file
> ![image](https://user-images.githubusercontent.com/48523533/99932835-13b83a00-2d7b-11eb-81b7-f26a73dc97de.png)

My config file matches with you.
Also same with me. Evaluates after 1000 steps.
I ran after training finishes. (serial manner) I will try parallel.
Also currently with logdir as /train I get train loss in Tensorboard and with /eval I get only eval mAP, I can't get eval loss.
So what logdir you used with tensorboard ? /train ?

**Comment by koolvn** [comment]
@adityap27 You _have to_ run eval task in parallel (I also tried to run eval task after the training is finished but it gave me results only for the last step)
Actually I use $MODEL_DIR as a --logdir in tensorboard
The /eval and /train are subfolders in $MODEL_DIR

**Comment by adityap27** [comment]
> val and /train are subf


**Thanks** @koolvn ,
Running eval job in parallel and using $MODEL_DIR as a --logdir in tensorboard. These both worked for me to get nice graphs..!!

**Comment by Rishav-hub** [comment]
@koolvn what was your size of the dataset. I am training my model in collab and getting a loss of **0.036** for the same configuration that is mentioned above and for **5000** steps. But it seems that when I am visualizing the output it's not able to detect and also draw the bounding box. Any help??

**Comment by koolvn** [comment]
@Rishav-hub sorry, I can't remember what was the size of my dataset. Probably five thousand steps is not enough. Try training for 50k steps. Also use validation dataset to monitor metrics, especially recall.

**Comment by Rishav-hub** [comment]
> @Rishav-hub sorry, I can't remember what was the size of my dataset. Probably five thousand steps is not enough. Try training for 50k steps. Also use validation dataset to monitor metrics, especially recall.

But the loss is not decreasing after 8K steps also it remains between 0.04 to 0.03.

**Comment by dademiller360** [comment]
I'm trying to train the EfficientDet d0 512x512, but it doesn't seem to converge.

batch_size: 4
learning_rate_base: 0.00249999994412064
total_steps: 300000
warmup_learning_rate: 0.0010000000474974513
warmup_steps: 2500

![image](https://user-images.githubusercontent.com/78317189/107188852-b41c1300-69e8-11eb-9175-a8eb21c623d1.png)

any tips?

**Comment by koolvn** [comment]
@dademiller360 sims like learning rate is too high for your batch size. Try dividing LR by 10 or 100

**Comment by dademiller360** [comment]
> @dademiller360 sims like learning rate is too high for your batch size. Try dividing LR by 10 or 100

Thank you koolvn
Dividing LR by 10 seems to solve it
![image](https://user-images.githubusercontent.com/78317189/107749916-8e11ae00-6d1b-11eb-9731-eedea3dd0b9e.png)

**Comment by BogoK** [comment]
> Try to reduce the learning rate, this should solve the problem. The models are usually trained with a large batchsize, eg.: 128 for efficient_det d3, when adjusting this parameter (eg: 128 -> 4) makes sure to also change the learning rate, since your gradients are a lot noisier. I would divide the learning rate by the same factor.

How do you know, that 128 is optimal for d3? How much images for d0? I have only 28 examples in my whole dataset, using batch_size=1 and learning_rate=0.001. The loss jumps a lot and doesn't converge at all

**Comment by koolvn** [comment]
> How do you know, that 128 is optimal for d3? How much images for d0? I have only 28 examples in my whole dataset, using batch_size=1 and learning_rate=0.001. The loss jumps a lot and doesn't converge at all

Usually we want to have the batch size as big as possible.
In your case loss jumps and algorithm doesn't converge because you have a very big learning rate with respect to your small batch size. Try a smaller LR, something like 1e-6 or even less. The other way is to increse the batch size and augment your data.

[Here's a good explanation](https://medium.com/mini-distill/effect-of-batch-size-on-training-dynamics-21c14f7a716e)

**Comment by BogoK** [comment]
> > How do you know, that 128 is optimal for d3? How much images for d0? I have only 28 examples in my whole dataset, using batch_size=1 and learning_rate=0.001. The loss jumps a lot and doesn't converge at all
> 
> Usually we want to have the batch size as big as possible.
> In your case loss jumps and algorithm doesn't converge because you have a very big learning rate with respect to your small batch size. Try a smaller LR, something like 1e-6 or even less. The other way is to increse the batch size and augment your data.
> 
> [Here's a good explanation](https://medium.com/mini-distill/effect-of-batch-size-on-training-dynamics-21c14f7a716e)

Thank you. If I go lover than 0.001 my mAP is 0 for each of tested learning rates.
Which effect does the "steps" parameter has on the training?

**Comment by lihe07** [comment]
Same problem and solved when LR is reduced to 1e-5.
My batch_size is 4 and it trains on 4 GPUs.

**Comment by jidd20118** [comment]
Check this tutorial:: [link](https://github.com/abdelrahman-gaber/tf2-object-detection-api-tutorial)
